integrators/symplectic_runge_kutta_nyström_integrator_body.hpp

```diff
@@ -0,0 +1,336 @@
+  DoublePrecision<double> c_i = 0.0;
+  for (int i = 0; i < stages_; ++i) {
+    c_[i] = c_i.value;
+    c_i.Increment(a_[i]);
+  }
+  CHECK_EQ(1.0, c_i.value);
+  if (composition == kABA) {
+    CHECK_EQ(0.0, b_[0]);
+  } else if (composition == kBAB) {
+    CHECK_EQ(0.0, a_[stages_ - 1]);
+  }
+  if (time_reversible) {
+    switch (composition) {
+    case kABA:
+      for (int i = 0; i < stages_; ++i) {
+        CHECK_EQ(a_[i], a_[stages_ - 1 - i]);
+        CHECK_EQ(b_[i + 1], b_[stages_ - 1 - i]);
+      }
+      break;
+    case kBAB:
+      for (int i = 0; i < stages_; ++i) {
+        CHECK_EQ(b_[i], a_[stages_ - 2 - i]);
+        CHECK_EQ(b_[i], b_[stages_ - 1 - i]);
+      }
+      break;
+    case kBA:
+      LOG(FATAL) << "Time-reversible compositions have the FSAL property";
+      break;
+    default:
+      LOG(FATAL) << "Invalid CompositionMethod";
+    }
+  }
+}
+
+template<typename Position, int order, bool time_reversible, int evaluations,
+         CompositionMethod composition>
+void SymplecticRungeKuttaNyströmIntegrator<Position, order, time_reversible,
+                                           evaluations, composition>::Solve(
+    IntegrationProblem<ODE> const& problem,
+    Time const& step) const {
+  using Displacement = typename ODE::Displacement;
+  using Velocity = typename ODE::Velocity;
+  using Acceleration = typename ODE::Acceleration;
+
+  // Argument checks.
+  CHECK_NOTNULL(problem.initial_state);
+  int const dimension = problem.initial_state->positions.size();
+  CHECK_EQ(dimension, problem.initial_state->velocities.size());
+  CHECK_NE(Time(), step);
+  Sign const integration_direction = Sign(step);
+  if (integration_direction.Positive()) {
+    // Integrating forward.
+    CHECK_LT(problem.initial_state->time.value, problem.t_final);
+  } else {
+    // Integrating backward.
+    CHECK_GT(problem.initial_state->time.value, problem.t_final);
+  }
+
+  typename ODE::SystemState current_state = *problem.initial_state;
+
+  // Time step.
+  Time h = step;
+  // Current time.  This is a non-const reference whose purpose is to make the
+  // equations more readable.
+  DoublePrecision<Instant>& t = current_state.time;
+
+  // Position increment.
+  std::vector<Displacement> ∆q(dimension);
+  // Velocity increment.
+  std::vector<Velocity> ∆v(dimension);
+  // Current position.  This is a non-const reference whose purpose is to make
+  // the equations more readable.
+  std::vector<DoublePrecision<Position>>& q = current_state.positions;
+  // Current velocity.  This is a non-const reference whose purpose is to make
+  // the equations more readable.
+  std::vector<DoublePrecision<Velocity>>& v = current_state.velocities;
+
+  // Current Runge-Kutta-Nyström stage.
+  std::vector<Position> q_stage(dimension);
+  // Accelerations at the current stage.
+  std::vector<Acceleration> g(dimension);
+
+  bool at_end = false;
+
+  for (;;) {
+    // TODO(egg): do we want to keep cover the whole integration interval, or
```

Comment:
We discussed that comment and came to the conclusion that the currently-implemented behaviour was fine anyway.